src/Action/Security/RefreshAccessTokenAction.php

```diff
@@ -15,85 +15,95 @@
+use OAT\SimpleRoster\Security\Generator\JwtTokenGenerator;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Log\LoggerInterface;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
-use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
-use Symfony\Component\HttpKernel\Exception\ConflictHttpException;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Throwable;
 
 class RefreshAccessTokenAction
 {
-    /** @var TokenGenerator $generator */
+    /** @var JwtTokenGenerator */
     private $generator;
 
-    /** @var TokenStorage $storage */
-    private $storage;
-
     /** @var UserRepository */
     private $userRepository;
 
-    /** @var JwtTokenVerifierInterface */
-    private $tokenVerifier;
+    /** @var CacheItemPoolInterface */
+    private $tokenCache;
+
+    /** @var JwtTokenCacheIdGenerator */
+    private $tokenCacheIdGenerator;
 
     /** @var SerializerResponder */
     private $responder;
 
+    /** @var LoggerInterface */
+    private $logger;
+
     /** @var int */
     private $accessTokenTtl;
 
     public function __construct(
-        TokenGenerator $generator,
-        TokenStorage $storage,
+        JwtTokenGenerator $generator,
         UserRepository $userRepository,
-        JwtTokenVerifierInterface $tokenVerifier,
+        CacheItemPoolInterface $jwtTokenCache,
+        JwtTokenCacheIdGenerator $jwtTokenCacheIdGenerator,
         SerializerResponder $responder,
+        LoggerInterface $securityLogger,
         int $jwtAccessTokenTtl
     ) {
         $this->generator = $generator;
-        $this->storage = $storage;
         $this->userRepository = $userRepository;
-        $this->tokenVerifier = $tokenVerifier;
+        $this->tokenCache = $jwtTokenCache;
+        $this->tokenCacheIdGenerator = $jwtTokenCacheIdGenerator;
         $this->responder = $responder;
+        $this->logger = $securityLogger;
         $this->accessTokenTtl = $jwtAccessTokenTtl;
     }
 
-    public function __invoke(Token $refreshToken): Response
+    public function __invoke(Request $request, Token $refreshToken): Response
     {
         try {
-            $username = $refreshToken->getClaim('username');
-        } catch (\Throwable $exception) {
-            throw new ConflictHttpException('Invalid token. User claim is missing');
+            $user = $this->userRepository->findByUsernameWithAssignments($refreshToken->getClaim('aud'));
+        } catch (Throwable $exception) {
+            throw new AccessDeniedHttpException('Invalid token.');
         }
 
-        $user = $this->userRepository->findOneByUsername($username);
-
-        if (is_null($user)) {
-            throw new ConflictHttpException('Invalid token. User not found');
-        }
-
-        $cachedToken = $this->storage->getStoredToken($username);
+        $cacheId = $this->tokenCacheIdGenerator->generate($refreshToken);
+        $refreshTokenCacheItem = $this->tokenCache->getItem($cacheId);
```

Comment:
We compare the tokens. If they are not matching, the token is considered as invalid. By storing the refresh tokens in cache we can invalidate them and prompting the users to re-authenticate in case we need it.

Reply:
I have one concern here. I have fear that we can increase the memory usage on the servers considering that usually JWT are big tokens. Is it an issue?
In case of yes I'd propose to store it in a different way. we'd store only a hash of invalid tokens with the same expiration as the token. With this approach we would check if the token was invalidated instead of to check if the token is valid and it'd check based on a hash (we wouldn't save the token state). I believe that we could decrease the memory usage significantly like this way.

Reply:
I see your concern but I don't see an issue here, it takes less space than storing the sessions as we are doing it now. I compared the size of a random symfony session vs a JWT refresh token:

Symfony session (2394 bytes):

`
"_sf2_attributes|a:1:{s:13:\"_security_api\";s:1930:\"C:74:\"Symfony\\Component\\Security\\Core\\Authentication\\Token\\UsernamePasswordToken\":1841:{a:3:{i:0;N;i:1;s:3:\"api\";i:2;a:5:{i:0;O:28:\"OAT\\SimpleRoster\\Entity\\User\":7:{s:32:\"\x00OAT\\SimpleRoster\\Entity\\User\x00id\";i:1;s:38:\"\x00OAT\\SimpleRoster\\Entity\\User\x00username\";s:11:\"test-user-1\";s:38:\"\x00OAT\\SimpleRoster\\Entity\\User\x00password\";s:96:\"$argon2id$v=19$m=1024,t=3,p=1$rPekEgybx+1c+6GouhUM3g$CknF+cqCLS8u+OYwFSKyCEUofD+CMDkV0hLpkUaHI6Y\";s:41:\"\x00OAT\\SimpleRoster\\Entity\\User\x00assignments\";O:33:\"Doctrine\\ORM\\PersistentCollection\":2:{s:13:\"\x00*\x00collection\";O:43:\"Doctrine\\Common\\Collections\\ArrayCollection\":1:{s:53:\"\x00Doctrine\\Common\\Collections\\ArrayCollection\x00elements\";a:1:{i:0;O:34:\"OAT\\SimpleRoster\\Entity\\Assignment\":7:{s:38:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00id\";i:1;s:41:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00state\";s:5:\"ready\";s:40:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00user\";r:6;s:44:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00lineItem\";O:47:\"Proxies\\__CG__\\OAT\\SimpleRoster\\Entity\\LineItem\":9:{s:17:\"__isInitialized__\";b:1;s:36:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00id\";i:10;s:39:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00label\";s:14:\"20EVA2PROFR_QA\";s:37:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00uri\";s:53:\"https://depp.fr#September20-Delivery_QA_20eva2_pro_fr\";s:38:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00slug\";s:14:\"20EVA2PROFR_QA\";s:41:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00startAt\";N;s:39:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00endAt\";N;s:45:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00maxAttempts\";i:0;s:42:\"\x00OAT\\SimpleRoster\\Entity\\LineItem\x00isActive\";b:1;}s:45:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00updatedAt\";N;s:49:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00attemptsCount\";i:0;s:46:\"\x00OAT\\SimpleRoster\\Entity\\Assignment\x00lineItemId\";i:10;}}}s:14:\"\x00*\x00initialized\";b:1;}s:35:\"\x00OAT\\SimpleRoster\\Entity\\User\x00roles\";a:0:{}s:43:\"\x00OAT\\SimpleRoster\\Entity\\User\x00plainPassword\";N;s:37:\"\x00OAT\\SimpleRoster\\Entity\\User\x00groupId\";s:0:\"\";}i:1;b:1;i:2;N;i:3;a:0:{}i:4;a:1:{i:0;s:9:\"ROLE_USER\";}}}}\";}_sf2_meta|a:3:{s:1:\"u\";i:1610554085;s:1:\"c\";i:1610554085;s:1:\"l\";s:1:\"0\";}"
`

Refresh token: (628 bytes):

`
eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJodHRwOlwvXC9zaW1wbGUtcm9zdGVyLmRvY2tlci5sb2NhbGhvc3QiLCJzdWIiOiJyZWZyZXNoVG9rZW4iLCJhdWQiOiJ0ZXN0LXVzZXItMSIsImp0aSI6Ijc1MmJhZDYwLTMxOWEtNDQ4Yy1hYzQxLTNhM2FlOTlhNzRlOCIsImlhdCI6MTYxMDQ4MDg5MiwibmJmIjoxNjEwNDgwODkyLCJleHAiOjE2MTA1NjcyOTJ9.BnkMGVVdT2j5HzK8-fbXPiJXTSHiAmnVudEhzmW0_nMazktoL0cDppFLyu6ITsFx9BDos6xAgHieWbe--3meA-9zgwyb4vlsbzZd66PItw0tLcSfzTSI9P2EWf_BBPw8mCzG0aCyOAPZHSMAtPKl6LmBsefa4rTlGfC2CQ0JPCzxSvVdNbyFQnU4Cpi3dzV8jfaS7gcyjOY-JNECo1WGkV14ddBoCCQfv9LLPtelwe6ixW8lH9vQ3gCq0iNUnFjJ4HQVWsWG4kXdSTuoHKYHyfZYhibs-n4nx8si3FYqaLHcpGnVTAubJOALunP9QMyrZwQBuuMZWf45fszINIk4kw
`

So I think on average we can say that we save ~75% space by storing refresh tokens instead of sessions. What do you think?

Reply:
Agree with Dani